rust/ql/lib/codeql/rust/internal/TypeInference.qll

```diff
@@ -1205,12 +1211,27 @@ private module MethodCall {
 
     pragma[nomagic]
     override Type getTypeAt(TypePath path) {
-      exists(TypePath path0 | result = inferType(super.getReceiver(), path0) |
-        path0.isCons(TRefTypeParameter(), path)
+      exists(TypePath path0, Type t0 |
+        t0 = inferType(super.getReceiver(), path0) and
+        (
+          path0.isCons(TRefTypeParameter(), path)
+          or
+          not path0.isCons(TRefTypeParameter(), _) and
+          not (path0.isEmpty() and result = TRefType()) and
+          path = path0
+        )
+      |
+        result = t0
         or
-        not path0.isCons(TRefTypeParameter(), _) and
-        not (path0.isEmpty() and result = TRefType()) and
-        path = path0
+        // We do not yet model the `Deref` trait, so we hard-code the fact that
+        // `String` dereferences to `str` here. This allows us e.g. to resolve
+        // `x.parse::<usize>()` to the function `<core::str>::parse` when `x` has
+        // type `String`.
+        //
+        // See also https://doc.rust-lang.org/reference/expressions/method-call-expr.html#r-expr.method.autoref-deref
+        path.isEmpty() and
+        t0.(StructType).asItemNode() instanceof StringStruct and
+        result.(StructType).asItemNode() instanceof Builtins::Str
```

Comment:
Is it OK that we also still get the result from `result = t0` as well in this situation, i.e. that we get two results (of which only one applies depending on context)?